File: src/main/java/com/example/adplatform/domain/model/Advertisement.java
package com.example.adplatform.domain.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

/**
 * Domain entity representing an Advertisement in the system.
 * This is a persistence-ignorant domain entity following DDD principles.
 * Advertisements can be sourced from storage or YouTube and have
 * targeting rules based on geolocation, biographical data, and user mood.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Advertisement {

    private Long id;
    private String title;
    private String description;
    private String content;
    private AdvertisementSource source;
    private String sourceIdentifier; // File path or YouTube ID
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private boolean active;

    // Targeting rules
    private Set<GeoTarget> geoTargets = new HashSet<>();
    private Set<BioTarget> bioTargets = new HashSet<>();
    private Set<MoodTarget> moodTargets = new HashSet<>();

    /**
     * Business method to check if this advertisement is valid for the given targeting criteria.
     */
    public boolean matchesTargeting(
            String countryCode,
            Integer age,
            String gender,
            String occupation,
            String educationLevel,
            String language,
            Set<String> interests,
            Mood mood,
            Integer intensity,
            String timeOfDay,
            String dayOfWeek,
            String season
    ) {
        return matchesGeoTargeting(countryCode, null, null, null, null)
                && matchesBioTargeting(age, gender, occupation, educationLevel, language, interests)
                && matchesMoodTargeting(mood, intensity, timeOfDay, dayOfWeek, season);
    }

    /**
     * Check if this advertisement matches the given geo targeting criteria.
     */
    public boolean matchesGeoTargeting(
            String countryCode,
            String region,
            String city,
            Double latitude,
            Double longitude
    ) {
        // If no geo targets are defined, the ad doesn't match specific geo targeting
        if (geoTargets == null || geoTargets.isEmpty()) {
            return false;
        }

        // Check for any matching include targets
        boolean hasIncludeMatch = geoTargets.stream()
                .filter(target -> target.isInclude())
                .anyMatch(target -> target.matches(countryCode, region, city, latitude, longitude));

        // Check for any matching exclude targets
        boolean hasExcludeMatch = geoTargets.stream()
                .filter(target -> !target.isInclude())
                .anyMatch(target -> target.matches(countryCode, region, city, latitude, longitude));

        // Ad matches if it has at least one include match and no exclude matches
        return hasIncludeMatch && !hasExcludeMatch;
    }

    /**
     * Check if this advertisement matches the given biographical targeting criteria.
     */
    public boolean matchesBioTargeting(
            Integer age,
            String genderStr,
            String occupation,
            String educationLevel,
            String language,
            Set<String> interests
    ) {
        // If no bio targets are defined, the ad doesn't match specific bio targeting
        if (bioTargets == null || bioTargets.isEmpty()) {
            return false;
        }

        final Gender gender = genderStr != null ? Gender.fromString(genderStr) : null;

        // Check for any matching include targets
        boolean hasIncludeMatch = bioTargets.stream()
                .filter(target -> target.isInclude())
                .anyMatch(target -> target.matches(age, gender, occupation, educationLevel, language, interests));

        // Check for any matching exclude targets
        boolean hasExcludeMatch = bioTargets.stream()
                .filter(target -> !target.isInclude())
                .anyMatch(target -> target.matches(age, gender, occupation, educationLevel, language, interests));

        // Ad matches if it has at least one include match and no exclude matches
        return hasIncludeMatch && !hasExcludeMatch;
    }

    /**
     * Check if this advertisement matches the given mood targeting criteria.
     */
    public boolean matchesMoodTargeting(
            Mood mood,
            Integer intensity,
            String timeOfDay,
            String dayOfWeek,
            String season
    ) {
        // If no mood targets are defined, the ad doesn't match specific mood targeting
        if (moodTargets == null || moodTargets.isEmpty()) {
            return false;
        }

        // Check for any matching include targets
        boolean hasIncludeMatch = moodTargets.stream()
                .filter(target -> target.isInclude())
                .anyMatch(target -> target.matches(mood, intensity, timeOfDay, dayOfWeek, season));

        // Check for any matching exclude targets
        boolean hasExcludeMatch = moodTargets.stream()
                .filter(target -> !target.isInclude())
                .anyMatch(target -> target.matches(mood, intensity, timeOfDay, dayOfWeek, season));

        // Ad matches if it has at least one include match and no exclude matches
        return hasIncludeMatch && !hasExcludeMatch;
    }
}

File: frontend/types/advertisement.ts
/**
 * Enum representing the source of an advertisement.
 */
export enum AdvertisementSource {
  STORAGE = 'STORAGE',
  YOUTUBE = 'YOUTUBE'
}

/**
 * Interface for geolocation targeting criteria.
 */
export interface GeoTarget {
  countryCode: string;
  region?: string;
  city?: string;
  latitude?: number;
  longitude?: number;
  radiusKm?: number;
  include: boolean;
}

/**
 * Enum representing gender options for targeting.
 */
export enum Gender {
  MALE = 'MALE',
  FEMALE = 'FEMALE',
  OTHER = 'OTHER',
  ALL = 'ALL'
}

/**
 * Interface for biographical targeting criteria.
 */
export interface BioTarget {
  minAge?: number;
  maxAge?: number;
  gender?: Gender;
  occupation?: string;
  educationLevel?: string;
  language?: string;
  interestCategory?: string;
  include: boolean;
}

/**
 * Enum representing different mood states for targeting.
 */
export enum Mood {
  HAPPY = 'HAPPY',
  SAD = 'SAD',
  EXCITED = 'EXCITED',
  RELAXED = 'RELAXED',
  STRESSED = 'STRESSED',
  BORED = 'BORED',
  FOCUSED = 'FOCUSED',
  TIRED = 'TIRED',
  ENERGETIC = 'ENERGETIC',
  NEUTRAL = 'NEUTRAL'
}

/**
 * Interface for mood-based targeting criteria.
 */
export interface MoodTarget {
  mood: Mood;
  intensityMin?: number;
  intensityMax?: number;
  timeOfDay?: string;
  dayOfWeek?: string;
  season?: string;
  include: boolean;
}

/**
 * Interface representing an Advertisement.
 */
export interface YouTubeDetails {
  videoId?: string;
  videoTitle?: string;
  channelId?: string;
  channelTitle?: string;
  durationSeconds?: number;
  thumbnailUrl?: string;
  publishedAt?: string;
}

export interface Advertisement {
  id?: number;
  title: string;
  description?: string;
  content: string;
  source: AdvertisementSource;
  sourceIdentifier: string;
  createdAt?: string;
  updatedAt?: string;
  active: boolean;
  // Click behavior
  targetUrl?: string;
  clickable?: boolean;
  youtubeDetails?: YouTubeDetails;
  geoTargets?: GeoTarget[];
  bioTargets?: BioTarget[];
  moodTargets?: MoodTarget[];
}

/**
 * Interface for user biographical data.
 */
export interface UserBioData {
  age?: number;
  gender?: string;
  occupation?: string;
  educationLevel?: string;
  language?: string;
  interests?: string[];
}

/**
 * Interface for targeting request parameters.
 */
export interface TargetingParams {
  countryCode?: string;
  region?: string;
  city?: string;
  latitude?: number;
  longitude?: number;
  userBioData?: UserBioData;
  mood?: Mood;
  intensity?: number;
  timeOfDay?: string;
  dayOfWeek?: string;
  season?: string;
}